urld0f681c558_farom57_AstroidDriver_tgz-pJ8-laazotea_indi_server_examples_MinimumINDIServerJ8/src/farom/astroiddriver/INDIAstroidDriver.java

```diff
@@ -0,0 +1,1636 @@
+				}
+			}
+
+			if (commandChanged) {
+				sendCommand();
+			}
+
+		}
+	}
+
+	/**
+	 * returns the modulus 24 in [0, 24[
+	 * 
+	 * @param x
+	 * @return x modulus 24 in [0, 24[
+	 */
+	private double mod24(double x) {
+		double res = x % 24;
+		if (res >= 0) {
+			return res;
+		} else {
+			return res + 24;
+		}
+	}
+
+	/**
+	 * returns the modulus 360 in [-180, 180[
+	 * 
+	 * @param x
+	 * @return x modulus 360 in [-180, 180[
+	 */
+	private double mod360(double x) {
+		double res = x % 360;
+		res += (res < 0 ? 360 : 0);
+		res += (res >= 180 ? -360 : 0);
+		return res;
+	}
+	
+	/**
+	 * Execute the focus servo move
+	 * @param time duration of the move in s
+	 * @param speed negative or positive speed
+	 */
+	private void moveFocus(final double duration, final double speed, final INDIProperty prop){
+		updateTicks((int) (neutralTicksE.getValue() + speed));
+		TimerTask task = new TimerTask(){
+			@Override
+			public void run() {
+				updateTicks(CmdMessage.TICKS_OFF);
+				prop.setState(PropertyStates.OK);
+				absFocusPosE.setValue(absFocusPosE.getValue()+speed*duration);
+				try {
+					updateProperty(prop);
+					updateProperty(absFocusPosP);
+				} catch (INDIException e) {
+					e.printStackTrace();
+				}
+			}			
+		};
+		Timer timer = new Timer();
+		timer.schedule(task, (long) (duration*1000));
+		
+		prop.setState(PropertyStates.IDLE);
+		try {
+			updateProperty(focusTimerP);
+		} catch (INDIException e) {
+			e.printStackTrace();
+		}
+		
+	}
+	
+	private void activeFocuser(){
+		addProperty(focusMotionP);
+		addProperty(focusSpeedP);
+		addProperty(focusTimerP);
+		addProperty(relFocusPosP);
+		addProperty(absFocusPosP);
+	}
+	
+	private void disableFocuser(){
+		removeProperty(focusMotionP);
+		removeProperty(focusSpeedP);
+		removeProperty(focusTimerP);
+		removeProperty(relFocusPosP);
+		removeProperty(absFocusPosP);
+	}
+	
+	private void activeIntervalometer(){
+		addProperty(intervalometerSettingsP);
+		initIntervalometer();
+	}
+	
+	private void disableIntervalometer(){
+		removeProperty(intervalometerSettingsP);
+		stopIntervalometer();
+	}
+	
+
+	Timer intervalometer;
+	
+
+	TimerTask currentTask;	
+	
+
+	
+	private void initIntervalometer(){
+		intervalometer = new Timer();
+		
+		
+	}
+	
+	private void startIntervalometer(){
+		printMessage("Start");
+		currentTask = new CompleteTask();
+		updateTicks(CmdMessage.TICKS_OFF);
+
+		intervalometerSettingsP.setState(PropertyStates.BUSY);
+		try {
+			updateProperty(intervalometerSettingsP);
+		} catch (INDIException e) {
+			e.printStackTrace();
+		}
+		
+		double n = exposureNumberE.getValue();
+		if(n>0){
+			if(mirrorRaisingTimeE.getValue()>0.1){
+				intervalometer.schedule(new MirrorRaisingTask(), 100);
+			}else{
+				intervalometer.schedule(new ExposeTask(), 100);
+			}
+		}else{
+			intervalometerSettingsP.setState(PropertyStates.OK);
+			try {
+				updateProperty(intervalometerSettingsP);
+			} catch (INDIException e) {
+				e.printStackTrace();
+			}
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 1526:1533, 1622:1629.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.